src/Device/Driver/AR62xx.cpp

```diff
@@ -0,0 +1,670 @@
+  virtual bool DataReceived(const void *data, size_t length, struct NMEAInfo &info) override;
+};
+
+// Little helpers for me
+//size_t = long unsigned int
+//unit16_t = unsigned short int
+//uint8_t = unsigned char
+//TCHAR = char
+
+/*
+ * Converting to a hex string for a useful output via LogFormat
+ */
+uint8_t AR62xxDevice::toHexString(const void* data, size_t size) {
+  uint8_t szHex = 0;
+  constexpr TCHAR hex_chars[16] = {
+          _T('0'), _T('1'), _T('2'), _T('3'), _T('4'), _T('5'), _T('6'), _T('7'),
+          _T('8'), _T('9'), _T('A'), _T('B'), _T('C'), _T('D'), _T('E'), _T('F')
+  };
+  const uint8_t* p = static_cast < const uint8_t* > (data);
+  const uint8_t* pend = p + size;
+  for(;p < pend; ++p) {
+    szHex += (hex_chars[((*p) & 0xF0) >> 4]);
+    szHex += (hex_chars[((*p) & 0x0F) >> 0]);
+    szHex += ' ';
+  }
+  return szHex;
+}
+
+/*
+ * Constructor
+ * Port on which the radio is connected
+ */
+AR62xxDevice::AR62xxDevice(Port &_port) : port(_port) {
+  response = ACK;
+  RadioPara.Enabled8_33 = true;
+}
+
+/**
+ * Receives and handles data from the radio.
+ *
+ * The function parses messages send by the radio.
+ * Because all control characters (e.g. HEADER_ID, PROTOKOLL_ID, STX, ACK, NAK, ...)
+ * can be part of the payload of the messages, it is important
+ * to separate the messages to distinguish control characters
+ * from payload characters.
+ *
+ * If a response to a command is received, the function notifies
+ * the sender. This could trigger a retransmission in case of a
+ * failure.
+ *
+ * The initial frequency settings of the radio a delivered by this method and stored in the data struct "info"
+ * every time connection is established
+ */
+bool AR62xxDevice::DataReceived(const void *_data, size_t length, struct NMEAInfo &info) {
+  assert(_data != nullptr);
+  assert(length > 0);
+  const uint8_t *data = (const uint8_t *)_data;
+  bool doneok = AR620xParseString((const char *)data, length);
+  info.alive.Update(info.clock);
+  return doneok;
+}
+
+/*
+ * Writes the message to the serial port on which the radio is connected
+ */
+bool AR62xxDevice::Send(const uint8_t *msg, unsigned msg_size, OperationEnvironment &env) {
+  //! Number of tries to send a message
+  unsigned retries = NR_RETRIES; // i.e. 3 retries
+  assert(msg_size > 0);
+  do {
+    bSending = true;
+    response_mutex.lock();
+    response = NO_RSP;
+    response_mutex.unlock();
+    // Send the message
+    if (!port.FullWrite(msg, msg_size, env, std::chrono::seconds(CMD_TIMEOUT))){
+      bSending = false;
+      response = NAK;
+    } else {
+      response = ACK;
+    }
+    // Wait for the response
+    response_mutex.lock();
+    rx_cond.wait_for(response_mutex, std::chrono::seconds(CMD_TIMEOUT));
+    
+    auto _response = response;
+    response_mutex.unlock();
+    bSending = false;
+    //LogFormat(_T("AR620x AR62xxDevice::Send: gesendet = %d Zeichen= 0x%04X response= 0x%04X"), msg_size, toHexString(msg, msg_size), _response);
```

Comment:
Huh?

Reply:
Huhuh?

Reply:
Rephrased: why is this code line here and what does it mean?